### DISCUSSION TITLE
Stretch the top and bottom of an image

### DISCUSSION THREAD

**Discussion by SgiarC** [ImageMagick/ImageMagick#4406] [Digital Image Processing] [answered]
Hello,

I would like to make an image where the horizontal center is normal and the the top and the bottom are progressively stretched so the height would be taller. My end goal is to have an image put on a horizontal pipe, then when you look at it straight, the image looks like the flat original image.

I have fooled around with different types of -distort but I can't get anything close to what I'm trying to do. Is there some other command that would be more suited to what I'm trying to accomplish.

Thanks in advance for pointing me to some similar code or a general idea of what to try,
Craig

**Comment by snibgo** [comment]
For this, we need some maths. Specifically, we need the mathematical expression that, for each output pixel, tells is the location of the input pixel.

Your "horizontal pipe" is a cylinder, with a circular cross-section. Suppose its radius is "r". For simplicity, suppose we view the cylinder from an infinite distance, so we are not bothered about perspective. We can see only half the cylinder.

Suppose the input image is WxH pixels. Suppose the part of the cylinder is W pixels wide, and in the other direction the distance along the surface is "pi \* r" (because the circumference of a circle is 2\*pi\*r).

Before I go further, please check the above. Is this what you want?

If it is correct, the next stage is to state the projection. Imagine a plane resting on the cylinder that is W pixels wide and 2\*r pixels high. This is effectively what the viewer will see, and should look like the original WxH image, so 2\*r=H. We need the transformation from that plane to the surface of the cylinder.

**Comment by fmw42** [comment]
What version of Imagemagick and what platform? Is this what you want except horizontal?

Input:

![img5967](https://user-images.githubusercontent.com/13458218/138602783-6f275360-6df6-4a14-8b40-9ae7c6baa8bb.jpg)

Output:

![img5967_cyl_vertical_full_r28_l74_black_animation](https://user-images.githubusercontent.com/13458218/138602788-c402e3f8-5d45-45cd-9c1a-d3fb9ee04d91.gif)

If so, and you are on a Unix like system (Linux, Mac OSX or Windows 10 Unix), the I have a cylinderize script. You can also view in perspective. See my web site at http://www.fmwconcepts.com/imagemagick

User snibgo has some cylinderizing scripts also for Windows on his web site.

**Reply by SgiarC** [reply]
Fred, I went through a bunch of your examples, very nice work. I couldn't find one that did what is essentially the opposite of making a flat image look like a cylinder.
I am working on Linux and testing on the command line, but eventually I'll convert that into perl.

This page:
https://legacy.imagemagick.org/Usage/resize/
has an example that seems promising, but I couldn't get -liquid-rescale to work on my ImageMagick version, plus I couldn't figure out how to change the command to have it expand on the top and the bottom.
 convert logo_trimmed.jpg  -liquid-rescale 130x100%\!  logo_lqr_expand.jpg

![image](https://user-images.githubusercontent.com/93088734/138610807-1757a8d3-83e8-44af-92e8-468b1dc778e8.png)
![image](https://user-images.githubusercontent.com/93088734/138610823-22f8276d-4a28-4bf3-ab80-5ada814817c1.png)

Thanks, Craig

**Comment by snibgo** [comment]
I think that is the opposite of what the OP wants.

The common requirement is "make an image of what a cylinder would look like if a given image were wrapped around it".

But this requirement is (I think) "make an image that, when rolled into a physical cylinder, would then look like a given image".

**Reply by SgiarC** [reply]
snibgo, That's correct I would like the image when wrapped on a pipe to look like the original image, it's the opposite of what most examples show.

If it was a 3/4" pipe then the circumference would be about 2.36". So a good visible area would be 2.36/4 which is about .6 of an inch tall. If the original image was square my guess is that it would end up being 0.5" wide x 0.6" tall and when that was rolled onto a cylinder it would look square.

**Comment by snibgo** [comment]
The way I would approach this:

We assume that every cross-section of the cylinder is identical. So we need a displacement map that is 1xH pixels. When we know the maths, we can construct the map with "-fx", which will be fast enough because the map is small.

The maths is fairly simple geometry, but I don't have currently have enough time to figure it out. Sorry. Maybe in a day or so.

**Reply by SgiarC** [reply]
snibgo, I have also looked through your examples and they seem very good, but just to test something it's a little difficult for me to convert bat to just a simple command line code.

Here is a quick photoshop mock-up of a checker board where the top and bottom are stretched. 

![image](https://user-images.githubusercontent.com/93088734/138613862-414459e6-6fbc-49f6-91d8-c6a6f276b740.png)

I think something like this put on the face of a cylinder would look like a regular checker board when you looked straight at it..

**Reply by fmw42** [reply]
So what kind of mathematical stretch did photoshop do. You cannot just do any non-linear stretch and expect it to be correct for a cylinder. If you want your kind of indiscriminate stretch, see my xpand script.

**Reply by SgiarC** [reply]
Fred,
In photoshop I just manually modified the height of each row, just to print it out and see how it looked rolled up, also to make sure you knew what I meant by expanding the top and bottom.

I took a look at your xpand script. It's complex, flexible with good error checking, and a lot to digest. The output is very good, but for what I need it to do it's a little too slow. Overall for what I want to do, it doesn't need to be that precise.

I have a question about the -fx part, is it an ImageMagick loop of some kind and the loop creates sets of points?

==================
My input image:
cb.jpg is a 500x500 checker board pattern

Here is the core of your program for vertical (I added -verbose)

hoc=`convert xc: -format "%[fx:($dimension-10)/2]" info:`

convert -size ${wi}x${dimension} xc: $tmpI  \
-verbose -fx "v.p{ i, ( (j-$hoc) +
sign(j-$hoc)*pow((j-$hoc),2)*($hic-$hoc)/pow($hoc,2) + $hic)}" \
"$outfile"

Here is what I enter on the command line.
./xpand -d "100:85" -m vertical cb.jpg cb-expand-4.jpg

This is $hoc

convert -size 500x588 xc: ./XPAND_26425.miff

the hoc value is 293.5

I believe the -fx is some kind of an ImageMagick loop inside
a loop (i and j) which initially starts with:

-fx "v.p{ i, ( (j-$hoc) + sign(j-$hoc)*pow((j-$hoc),2)*($hic-$hoc)/pow($hoc,2) + $hic)}" \

-fx v.p{ i, ( (j-293.5) + sign(j-293.5)*pow((j-293.5),2)*(249.5-293.5)/pow(293.5,2) +249.5 )}

This is the verbose output
xc:=>cb-expand-4.jpg XC 500x588 500x588+0+0 16-bit sRGB 30063B 21.660u 0:07.662

=======================
Thanks, Craig

**Reply by fmw42** [reply]
%[fx:...] is a calculator. So it is a one time computation to get an variable for use later.

-fx processes the image pixel by pixel using the formula presented subsequently. But it is very slow as I used it to process every pixel, because it is not pre-compiled. Sometimes there are ways to use equivalent compiled methods to do similar things.

My function is using a power law expansion to do a non-linear expand. So it is exponential.

From my quick view of what snibgo provided, his is using a sin function as the inverse transform of the arcsin, which is likely the proper expression for wrapping on a cylinder. So I would suggest going with his implementation. His is fast because he uses -fx to compute a distortion map and then uses the compiled -distort method to do the processing.

**Comment by fmw42** [comment]
The cylinderize is a sin function. So I would assume (untested) that the opposite would be an arcsin function.

**Comment by snibgo** [comment]
I found some time. Well, it's past my bedtime. Windows BAT script:

```
set AMAX=90 * pi/180

set FMT=^
WW=%%[fx:w]\n^
AMAX=%%[fx:%AMAX%]\n^
HMAX=%%[fx:h/2]\n

for /F "usebackq" %%L in (`%IMG7%magick ^
  toesGrid.png ^
  -precision 15 ^
  -format "%FMT%" ^
  info:`) do set %%L

set FMT=^
RR=%%[fx:%HMAX%/sin(%AMAX%)]\n

for /F "usebackq" %%L in (`%IMG7%magick ^
  xc: ^
  -precision 15 ^
  -format "%FMT%" ^
  info:`) do set %%L

set FMT=^
DMAXx2=%%[fx:int(%RR%*%AMAX%*2+0.5)]\n

for /F "usebackq" %%L in (`%IMG7%magick ^
  xc: ^
  -precision 15 ^
  -format "%FMT%" ^
  info:`) do set %%L

%IMG7%magick ^
  -size %%[fx:%DMAXx2%]x1 xc: ^
  -fx "DD=i-w/2; HH=(%RR%*sin(DD/%RR%)+w/2+0.5)/w; HH" ^
+write w0.png ^
  -rotate 90 ^
  ( -size 1x%WW% gradient:Black-White -rotate -90 ) ^
  +swap ^
  ( -size 1x1 xc:gray(50%%) ) ^
  -scale "%WW%x%%[fx:%DMAXx2%]^!" ^
  -combine ^
  map2d.png

%IMG7%magick ^
  ( toesGrid.png ^
    -background #88f ^
    -gravity Center ^
    -extent %WW%x%%[fx:%DMAXx2%] ^
+write e.png ^
  ) ^
  map2d.png ^
  -compose Distort -define compose:args=0x%%[fx:h/2] -composite ^
  out.png
```
The script isn't optimized, and it writes debugging images w0.png and e.png.

The first line is:
```
set AMAX=90 * pi/180
```
This is half the angle of the portion of the pipe we can see, something between 0 and 90 degrees.

The input, toesGrid.png, is:
![toesGrid](https://user-images.githubusercontent.com/33812368/138623276-94e256e1-5b46-4ea9-a8ba-2c1d0a638778.png)

For 90 degrees, the result is:
![c90](https://user-images.githubusercontent.com/33812368/138623167-c5aa01f8-0351-4948-b603-bdfe2e97737d.png)

For 60 degrees, the result is:
![c60](https://user-images.githubusercontent.com/33812368/138623194-7608ae9f-3ac2-4469-ba91-21735d89f4b8.png)

For 45 degrees, the result is:
![c45](https://user-images.githubusercontent.com/33812368/138623222-d8f44807-5e08-422f-a816-17409323f5b0.png)

**Comment by snibgo** [comment, marked as answer]
Here is a bash version. After doing the calculations (which I expect could be done in perl), there is only one IM command.
```
# Distort the input so that,
# when printed and wrapped around a pipe,
# that would look like the input image.

# $1 is input file
# $2 is output file
# $3 is half the angle of the pipe we can see, in degrees, 0 to 90.

infile=$1
outfile=$2
amaxDeg=$3

ww=$(magick $infile -precision 15 -format "%[w]" info: )

hmax=$(magick $infile -precision 15 -format "%[fx:h/2]" info: )

amax=$(magick xc: -precision 15 -format "%[fx:$amaxDeg * pi/180]" info: )

rr=$(magick xc: -precision 15 -format "%[fx:$hmax / sin($amax)]" info: )

dx2=$(magick xc: -precision 15 -format "%[fx:int($rr * $amax * 2 + 0.5)]" info: )

magick \
  \( $infile \
     -background "#88f" \
     -gravity Center \
     -extent ${ww}x${dx2} \
  \) \
  \( \
     -size ${dx2}x1 xc: \
     -fx "DD=i-w/2; HH=($rr *sin(DD/$rr)+w/2+0.5)/w; HH" \
     -rotate 90 \
     \( -size 1x$ww gradient:Black-White -rotate -90 \) \
     +swap \
     \( -size 1x1 xc:"gray(50%)" \) \
     -scale "${ww}x${dx2}!" \
     -combine \
  \) \
  -compose Distort -composite \
  $outfile
```
EDIT: Corrected minor mistake in calculation of amax.

**Reply by SgiarC** [reply]
snibgo,
It's much easier for me to digest the bash version, plus I can easily test it out. To get it to run I needed to remove all of the \r and change magick to convert.

Wow, the output is great and it's very quick. Here are a few of my test results.

--Original Image
![image](https://user-images.githubusercontent.com/93088734/138731272-e8967f04-e1f4-48ad-919a-8199ca066440.png)

-- at 45 degrees
![image](https://user-images.githubusercontent.com/93088734/138731457-385828da-3215-4c9b-a64f-fac7bccc7352.png)

-- at 50 degrees
![image](https://user-images.githubusercontent.com/93088734/138731551-843456eb-9a5a-480b-a09a-6832b55dbcb6.png)

-- at 55 degrees
![image](https://user-images.githubusercontent.com/93088734/138731736-9de8871e-c5d4-400e-bda4-82d27952c079.png)

Many Thanks,
Craig

PS Later today when I have time,  I would like to ask a couple of questions about the code.

**Reply by SgiarC** [reply]
Snibgo,

The output of your code is really good and it's quick. I want to convert this to perlMagick which would keep me from opening and saving the image externally, but some of the command line seems difficult to port into perlMagick code. Could I ask you a couple questions about the command line code? I put my guesses below.

Here are some values:

ww=128 hmax=64 amax=1.39626340159546 rr=64.9873031606877 dx2=181

-------------------------------

convert \

  \( $infile \
     -background "#88f" \
     -gravity Center \
     -extent 128x181 \
  \) \

The part above sets up the original image #one?

---------------------------------------

  \( \
     -size 181x1 xc: \
     -fx "DD=i-w/2; HH=($rr *sin(DD/$rr)+w/2+0.5)/w; HH" \
     -rotate 90 \
     \( -size 1x128 gradient:Black-White -rotate -90 \) \
     +swap \
     \( -size 1x1 xc:"gray(50%)" \) \
     -scale "128x181!" \
     -combine \
  \) \

The above part is creating a new canvas using xc:

My guess is that -fx works horizontally and the formula applies a width to each row, then you rotate so that it's vertical

This line within parens :
( -size 1x128 gradient:Black-White -rotate -90 \)

is that being applied to the xc: that has the -fx?

Then you have a +swap and then another 1x1 xc:

Then those two or three xc: are combined into one image #two?

--------------------------------------------

This next part combines those two images by Distorting the
original #one using #two and then saves the image?
  -compose Distort -composite \

  $outfile

----------------------------------------------

Thanks,
Craig

**Comment by snibgo** [comment]
Oops, there is a minor mistake. Change this line:
```
amax=$(magick xc: -precision 15 -format "%[fx:$3 * pi/180]" info: )
```
... to:
```
amax=$(magick xc: -precision 15 -format "%[fx:$amaxDeg * pi/180]" info: )
```

**Comment by snibgo** [comment]
I have tidied my script somewhat. It currently is:
```
#! /bin/bash

# pipe2.sh
# Last edited: 25-October-2021

# Distort the input so that,
# when printed and wrapped around a pipe,
# that would look like the input image.

# $1 is input file
# $2 is output file
# $3 is half the angle of the pipe we can see, in degrees, 0 to 90.

infile=$1
outfile=$2
amaxDeg=$3

ww=$(magick $infile -precision 15 -format "%[w]" info: )

w_2=$(magick $infile -precision 15 -format "%[fx:w/2]" info: )

hmax=$(magick $infile -precision 15 -format "%[fx:h/2]" info: )

amax=$(magick xc: -precision 15 -format "%[fx:$amaxDeg * pi/180]" info: )

rr=$(magick xc: -precision 15 -format "%[fx:$hmax / sin($amax)]" info: )

dx=$(magick xc: -precision 15 -format "%[fx:int($rr * ($amax) + 0.5)]" info: )

dx2=$(magick xc: -precision 15 -format "%[fx:int($rr * ($amax) * 2 + 0.5)]" info: )

echo ww=$ww w_2=$w_2 hmax=$hmax amax=$amax rr=$rr dx2=$dx2

magick \
  \( $infile \
     -background "#88f" \
     -gravity Center \
     -extent ${ww}x${dx2} \
  \) \
  \( \
     -size ${dx2}x1 xc: \
     -fx "($rr *sin((i-${dx})/$rr)+${dx}+0.5)/w" \
     -rotate 90 \
     \( -size 1x$ww gradient:Black-White -rotate -90 \) \
     +swap \
     \( -size 1x1 xc:"gray(50%)" \) \
     -scale "${ww}x${dx2}!" \
     -combine \
  \) \
  -compose Distort -composite \
  $outfile
```
A given angle $3 of 0 (zero) should give no distortion, but sin(0)=0 and we get division by zero. So add a check for that, or use a small angle such as 0.01 degrees.

The magick command creates one image in the first parentheses, and a second image (called an "absolute distortion map" or "absolute displacement map") in the second parens. Finally we apply the map to the first image, distorting the position of pixels.

For the distortion, we want the two images to be the same size, the required size of the output. $infile is shorter than this, so we make it taller with "-extent".

In your example, the $infile is 128 pixels tall, and we extend it to be 181 pixels tall.

> The part above sets up the original image #one?

Yes.

The next part creates the map, which will have three channels (R, G and B), 128x181 pixels.

First, we create the G channel, which will distort vertically. We make this 181x1 pixels then rotate so it is 1x181 pixels. Why? Because IM works best in rows, so it will process a 181x1 image more quickly than a 1x181 image.

I have improved the "-fx" operation to make it faster. This calculates a value for every pixel in that 181x1 image. For info on "-fx", see http://imagemagick.org/script/fx.php

"( -size 1x128 gradient:Black-White -rotate -90 ) " starts a new image. This will be the R channel, which will distort horizontally. It is a simple gradient because we want no distortion in the x-direction.

We want the R channel first and then the G channel, so we need to "+swap" them. Why didn't I create them in the correct order? I don't know. I'll change that.

Then we create the B channel. All the blue values will be the same, so we start with a single pixel, set to a mid gray.

Okay, now we have three images for the three channels. The sizes are 1x181 and 128x1 and 1x1. We "-scale" all three so they all become 128x181 pixels. Scaling is fast.

"-combine" makes one RGB image from the three inputs. So now we have the required distortion map.

Does that explain it? Do ask any questions.

**Reply by SgiarC** [reply]
Thanks for the thorough explanation.

Using perlMagick, I think it might be difficult to create the second image "absolute distortion map".

In the end I think I'll always be using the same size image, just the content of image # one would change. What do you think of the idea of saving the "absolute distortion map", possibly as a .miff file. 

Then in perlMagic, create image #one, then load the second .miff image and then with perlMagick use a variation of 
-compose Distort -composite
to distort each new image # one. 

If that's possible it would be super quick.

Thanks,
Craig

**Comment by snibgo** [comment]
> In the end I think I'll always be using the same size image, just the content of image # one would change. What do you think of the idea of saving the "absolute distortion map", possibly as a .miff file.

Yes, certainly. Just after "-combine", you can insert something like "+write mymap.miff". Then you can use that for any same-size inputs that need the same angle.

**Reply by SgiarC** [reply]
snibgo,

I used your +write suggestion and saved the file. Then after fussing around a little, this is what I came up with. I'm not sure if this is the best way to do it, but It's super quick and the image is accurate. 

convert \\
  \\( $infile \\
     -background "#88f" \\
     -gravity Center \\
     -extent ${ww}x${dx2} \\
  \\) \\
  \\( \\
     -size "${ww}x${dx2}" "mymap.miff" \\
  \\) \\
  -compose Distort -composite \\
  $outfile

Now I need to fuss around with perlMagick. I think the top part of putting the images together will be pretty simple, but I'll have to play around with the 
 -compose Distort -composite \
line a little bit.

Thanks,
Craig

**Comment by snibgo** [comment]
PS. I am curious. Can you say something about what you are using this for? If you can't say for commercial or other reasons, that's fine.

**Reply by SgiarC** [reply]
snibgo,

Let me work on it a little bit longer and I'll send you a sample image. Is there a way to send it to you directly without posting on the forum?

**Comment by snibgo** [comment]
Miff files contain metadata about their size, so you don't need "-size".

The input image needs to be extended to the size of the miff, so we don't need ${dx}. We read the miff first so we know its size, then extend $infile, and swap them so the order is correct for the "-composite". Like this:
```
magick \
  mymap.miff \
  -set option:MYSIZE %[fx:w]x%[fx:h] \
  \( \
     $infile \
     -background "#88f" \
     -gravity Center \
     -extent %[option:MYSIZE] \
  \) \
  +swap \
  -compose Distort -composite \
  $outfile
```

**Reply by SgiarC** [reply]
Thanks for the updated bash file.

I had noticed that the .miff files included a lot of metadata and it defined the w and h.
Although this line from my version of the bash file:
     -size "${ww}x${dx2}" "mymap.miff" \
is redundant, it was the only way I could find to get ImageMagick to accept the .miff file

When I give your updated bash file a spin, IM gets confused about the .miff file and throws these errors. (which was the problem I was always having until I added the -size)

By default, the image format of `file' is determined by its magic
number.  To specify a particular image format, precede the filename
with an image format name and a colon (i.e. ps:image) or specify the
image type as the filename suffix (i.e. image.ps).  Specify 'file' as
'-' for standard input or output.
./pipedistort: line 30: mymap.miff: command not found
./pipedistort: line 31: -set: command not found
./pipedistort: line 32: (: command not found
./pipedistort: line 33: mandrill_grid.jpg: command not found
./pipedistort: line 34: -background: command not found
./pipedistort: line 35: -gravity: command not found
./pipedistort: line 36: -extent: command not found
./pipedistort: line 37: ): command not found
./pipedistort: line 38: +swap: command not found
./pipedistort: line 39: -compose: command not found
./pipedistort: line 40: $'usemap-65.jpg\r': command not found

**Comment by snibgo** [comment]
> ./pipedistort: line 30: mymap.miff: command not found

The bash script interpreter is trying to interpret "mymap.miff" as a command, not as part of the "magick" command. Likewise "-set" and so on. Something is wrong with the line continuation characters, which should tell the interpreter that the "magick" command continues on the next line. Perhaps your command file has "\r\n" for new lines, but the interpreter accepts only "\n".

I use Windows. Text editors may use "\r\n" or "\n" only. Windows command interpreters accept either. I have setup bash on my computer to accept either. But many *nix computers will accept only "\n".

**Reply by SgiarC** [reply]
snibgo,
I suspected it was a \r problem, but when I tried to strip them out using my editor, it didn't find any. Something must be amiss with my editor as of today.

So for now I just run this command before I try the bash file:
sed -i 's/\r$//' usemap

**Reply by SgiarC** [reply]
Using this command:
  -set option:MYSIZE %[fx:w]x%[fx:h] \
is a nice option because I won't need to pass those values to the bash file.

When I try to use
-extent "%[MYSIZE]" \
or
-extent %[MYSIZE] \
or
-extent :"%[MYSIZE]" \

I get this error:
convert: invalid argument for option `-extent':
%[option:MYSIZE] @ error/convert.c/ConvertImageCommand/1497.

If I test the MYSIZE using
-annotate 0 "%[MYSIZE]" \
it writes the 128x153 correctly so I know the set is correct, it's something with -extent baulking with the %[MYSIZE].

**Reply by fmw42** [reply]
You can only use that syntax on IM 7 and when using magick, not convert. What is your IM version?

**Reply by SgiarC** [reply]
No wonder.
Version: ImageMagick 6.9.10-68

**Comment by SgiarC** [comment]
Since I am using perl this is how I convert the image without using a bash file.
$path is the complete path
$$ is the process ID
Since I am using 50% I have already generated the map "mapcode50.miff" and I know the size is 330x367
The system call below is all in one line

`system(qq{convert
"$path/mapcode50.miff"
\\(
"$path/code$$.jpg"
-background "#88f" -gravity Center -extent "330x367" \\)
+swap -compose Distort -composite
"$path/newcode$$.jpg"});`

I would like to figure out how to run that code using perlMagick without a system call.

**Comment by snibgo** [comment]
Sorry, I know nothing about perl.

I have improved the bash script, and written a Windows BAT version, and explained the maths: [Wrapping a pipe](http://im.snibgo.com/wrappipe.htm).

**Reply by SgiarC** [reply]
Snibgo,
Thanks for all your work on this. I'm still using your original version. 
-fx "DD=i-w/2; HH=($rr *sin(DD/$rr)+w/2+0.5)/w; HH" \\

When I tried to use one of your updated versions
-fx "($rr *sin((i-${dx})/$rr)+${dx}+0.5)/w" \\
something was amiss because it only created 1/2 the image. Most likely it was some typo that I had entered, so I went back to your original one. 

I'll take a look at your new Bat version and see if I can figure out how to port it to a bash file.

A few days ago I sent you an e-mail to your earthling.net address so you could see one of my project's test images. Perhaps it went to your junk mail.

I just put that perl line in there in case someone was curious about how to do it. For me usually it works better to call ImageMagick directly without going through a bash file. But, I do like testing using a bash file for testing.

At some point I might post a question about porting the ImageMagick script to perlMagick because it would eliminate one file write and just be a tad quicker, but using the "absolute distortion map", it's super quick anyway.

Thanks Again,
Craig

**Reply by SgiarC** [reply]
Snibgo,
Just had a chance to take a look at your new "Wrapping a Pipe" page. Wow that's really nice work and the animation is quite interesting. 
Craig

**Comment by snibgo** [comment]
Got the email, thanks.

From my web page, you will see that I suppose we are given the image height and the maximum angle, 0 to 90 degrees. From those the script calculates "rr", the pipe radius. Instead of the angle, we might be given the pipe radius (in pixels!) and the script can then calculate the maximum angle.

When the maximum angle is 90 degrees, there is a lot of stretch at the top and bottom. I expect angles up to 60 degrees should be fine.

My script assumes orthographic projection. This means that an object appears to be the same size however far away it is. If your eyeball (or camera) is close to the pipe, then the image near the top of the pipe will appear smaller than the image at the centre. We can correct for that, by widening the top and bottom to compensate.

Anyhow, the trick is to try it and see.

**Comment by snibgo** [comment]
I have extended my [Wrapping a pipe](http://im.snibgo.com/wrappipe.htm) page to include rectilinear projection, such as a camera held at a known height above the pipe. This gives a full perspective view, which modifies the vertical distortion and adds a horizontal distortion. The page includes a bash script for that distortion.
